text/0052-resource-importing-support.md

```diff
@@ -0,0 +1,268 @@
+solution:
+
+#### Inconsistent resource identifiers between resource types
+
+In the ResourceToImport structure, each type of resource is identified by a set
+of resource properties (typically of length 1 - by a single property). The
+property name is different between different resource types (e.g. `BucketName`
+for S3 buckets, or `VpcId` for VPCs).
+
+To be able to put the structure together, the import logic must be able to
+identify the required properties and then look up the values. For some
+resources, this can be automatically looked up from the CDK construct
+configuration (e.g. an S3 bucket having specific `bucketName` configured) but
+for some resources, the code has to rely on user input.
+
+To address this, the solution is divided into two steps:
+
+##### Physical resource identification
+
+First, identify what properties must be looked up/asked for for each resource to
+be imported. This information can be obtained by executing CloudFormation API
+action `GetTemplateSummary` with the new template. The action returns the
+property names used as "import identifiers" for each resource type used in the
+template.
+
+For each new resource (one that isn't present in the currently deployed
+template), look up the values for the "import identifiers". First, look up the
+resource (CDK construct) configuration. If not explicitly defined, interactively
+prompt user to provide the value.
+
+Based on command line flags, the compiled "resource mapping" structure is either
+passed on to the next phase, or written to the specified file.
+
+##### Import execution
+
+Based on command line flags, this phase either reads "resource mapping" from the
+specified file, or it will be passed the resource mapping directly from the
+previous phase.
+
+First, the sanity checks are performed and then the change set is created. By
+default, the change set is also executed (can be changed by passing
+`--no-execute` flag just like with regular deploy).
+
+#### Cannot import and create/change resources in the same changeset
+
+CloudFormation API doesn't allow any non-import changes in the template for
+import operation. This is a challenge for CDK, as:
+
+* `CDKMetadata` resource changes on each synthesis, so the import operation
+  never succeeds with the unmodified template. To resolve this, the proposed
+  solution restores CDKMetadata to the currently deployed version before
+  passing the template to ChangeSet creation step.
+* High-level constructs produce multiple underlying CloudFormation resources
+  and only some of those might need to be imported - the rest might need to
+  be created (e.g. `ec2.Vpc()` might require VPC being imported, but subnets
+  being created). Furthermore, CDK can't automatically determine what
+  resources to import and what to create. For this to behave correctly, the
+  solution will prompt user for each resource - whether it should be imported
+  or not. If not, the resource is removed from the template before creating
```

Comment:
Just thinking: the other way around might be easier.

Why don't we download the *current template* from the Stack, and then add the resources we're trying to import? Seems safer than dropping resources from the target template.

So the workflow looks a bit like:

* Take the new template, identify candidate resources to import by looking at `cdk diff`
* Obtain the import map by a combination of heuristics and user input. Include the resource definitions of the resources in the import map

Then:

* Download the old template
* Add the resources from the previous step to the template
* (Potentially: check the integrity of the template)
* Do an IMPORT update with the import map

Reply:
I wonder what we're going to do about dependencies on non-imported resources here... but I also don't have a good intuition for where and how those are going to crop up. I guess the best path forward is roll this feature out and seeing where it's going to creak 😅

Reply:
Yes, I am happy to revert the logic here, although it should produce exactly the same template, as:
* we already download the current template and perform the diff
* for any modification (besides CDKMetadata that we copy over) we fail so non-addition difference will never get to the import action
* for any newly defined resource, we decide whether it is going to be imported (we keep it) or not (we remove it) ... which would be the same other way around: if it's going to be imported we add it, otherwise we skip it

But now thinking about it, the above applies only to the `Resources` section of the template. I will double-check what if `Parameters` or `Outputs` change - this can add some more complexity. Imagine a resource properties of a resource-to-be-imported reference a parameter that did not exist in the previous template.

Reply:
Exactly. It just feels to me that taking a known-good template and doing a very concrete set of operations to it that we thought about beforehand is going to be safer than taking an unknown, new template, and hoping we remember to take out all the unintended modifications.

On the other hand, some `Parameters` and `Mappings` might be necessary to support the new resource?

Reply:
rephrased to reflect this